src/query/storage/m3/storage.go

```diff
@@ -329,58 +329,103 @@ func (s *m3storage) CompleteTags(
 	default:
 	}
 
-	// TODO: instead of aggregating locally, have the DB aggregate it before
-	// sending results back.
 	fetchQuery := &storage.FetchQuery{
 		TagMatchers: query.TagMatchers,
-		// NB: complete tags matches every tag from the start of time until now
-		Start: time.Time{},
-		End:   time.Now(),
 	}
 
-	results, cleanup, err := s.SearchCompressed(ctx, fetchQuery, options)
-	defer cleanup()
+	m3query, err := storage.FetchQueryToM3Query(fetchQuery, s.conversionCache)
 	if err != nil {
 		return nil, err
 	}
 
-	accumulatedTags := storage.NewCompleteTagsResultBuilder(query.CompleteNameOnly)
-	// only filter if there are tags to filter on.
-	filtering := len(query.FilterNameTags) > 0
-	for _, elem := range results {
-		it := elem.Iter
-		tags := make([]storage.CompletedTag, 0, it.Len())
-		for i := 0; it.Next(); i++ {
-			tag := it.Current()
-			name := tag.Name.Bytes()
-			if filtering {
-				found := false
-				for _, filterName := range query.FilterNameTags {
-					if bytes.Equal(filterName, name) {
-						found = true
-						break
-					}
+	aggOpts := storage.FetchOptionsToAggregateOptions(options, query)
+
+	var (
+		namespaces      = s.clusters.ClusterNamespaces()
+		accumulatedTags = storage.NewCompleteTagsResultBuilder(query.CompleteNameOnly)
+		multiErr        syncMultiErrs
+		wg              sync.WaitGroup
+	)
+
+	if len(namespaces) == 0 {
+		return nil, errNoNamespacesConfigured
+	}
+
+	// TODO: (arnikola): implement a better lifecycle for incoming aggregated tag
+	// iterators.
+	var mu sync.Mutex
+	aggIterators := make([]client.AggregatedTagsIterator, 0, initSize)
+	defer func() {
+		for _, it := range aggIterators {
+			it.Finalize()
+		}
+	}()
+
+	wg.Add(len(namespaces))
+	for _, namespace := range namespaces {
+		namespace := namespace // Capture var
+		go func() {
+			defer wg.Done()
+			session := namespace.Session()
+			namespaceID := namespace.NamespaceID()
+			aggTagIter, _, err := session.Aggregate(namespaceID, m3query, aggOpts)
+			if err != nil {
+				multiErr.add(err)
+				return
+			}
+
+			mu.Lock()
+			aggIterators = append(aggIterators, aggTagIter)
+			mu.Unlock()
+
+			completedTags := make([]storage.CompletedTag, aggTagIter.Remaining())
+			for i := 0; aggTagIter.Next(); i++ {
+				name, values := aggTagIter.Current()
+				tagValues := make([][]byte, values.Remaining())
+				for j := 0; values.Next(); j++ {
+					// TODO: (arnikola) fix the lifecycle here so we don't have to clone
```

Comment:
this feels like a redundant copy. you're making a local copy into `storage.CompletedTag` and then convert to `models.Metrics` above. can you do both in a single pass?

Reply:
Yeah, looking at the tag aggregator again looks like we cast to string, so the copy is unnecessary. Likewise the lifecycle is a bit less janky then I was expecting; there's definitely work to do on pooling/lifecycle management in m3query but that's pretty universally true rather than just in this case